### PR TITLE
Simplify trait bounds in dbsp.

### PR DESCRIPTION
**Pull request by gz**
We add the DBWeight+DBData trait throughout the layer API, this prepares for storage because it ensures that the types in the layers are serializable.

It also makes trait changes in the future less painful (hopefully) as we only have to change the DBWeight and DBData traits.

Is this a user-visible change (yes/no): no

### CODE REVIEW

**Comment by gz** [comment]
For review:

 - [ ] I adjusted the `seek_comparison_safety` test by changing how `RandomlyOrdered` works (making it thread-safe). I didn't fully understand what the test does so maybe checks that my changes preserve the original meaning

**Comment by gz** [comment]
mh surprisingly the CI passed, so I guess @ryzhyk you can review this now

File: crates/dbsp/src/trace/layers/column_layer/tests.rs [hunk 1]
```diff
@@ -52,18 +75,18 @@ impl<T> Item<T> {
 
 impl<T> PartialEq for Item<T>
 where
-    T: PartialEq,
+    T: DBData + DBWeight + PartialEq,
```
Comment (ryzhyk) [inline crates/dbsp/src/trace/layers/column_layer/tests.rs:78]: Just `DBWeight` should be sufficient as a trait bound here.

File: crates/dbsp/src/trace/layers/column_layer/tests.rs [hunk 2]
```diff
@@ -9,39 +9,62 @@ use crate::{
...
+        *ttl += 1;
     }
 }
 
-#[derive(Clone)]
-struct Item<T> {
+#[derive(Clone, Debug, Archive, Serialize, Deserialize, SizeOf)]
+struct Item<T>
+where
+    T: DBData + DBWeight,
```
Comment (ryzhyk) [inline crates/dbsp/src/trace/layers/column_layer/tests.rs:47]: `DBDatat + DBWeight` here and in other places should be redundant since `DBWeight: DBData`

**Review by ryzhyk — APPROVED** [review]

**Pull request merged**